ocrd_eval.md

```diff
@@ -0,0 +1,372 @@
+
+##### Example
+
+Given a Ground truth that reads `ſind` and the recognized text `fmd`.
+
+The Levenshtein distance between these texts is 4, because 4 edit operations are necessary to turn `fmd` into `ſind`:
+
+* `fmd` --> `ſmd` (substitution)
+* `ſmd` --> `ſimd` (insertion)
+* `ſimd` --> `ſind` (substitution)
+
+#### CER and WER
+
+##### Characters
+
+A text consists of a set of characters that have a certain meaning. A character is a glyph that represents a word, a letter in a word, or a symbol.
+Not included in the character definition are all forms of white spaces.
+
+###### Examples
+
+* the character `a` in the text `babst` represents the German letter `a`
+* the character `&` represents the Latin abbreviation `etc.`
+* the character `☿` represents an Astronomical symbol for the planet Mercury
+
+##### Character Error Rate (CER)
+
+The character error rate (CER) describes how many faulty characters the output of an OCR engine contains compaired to the Ground Truth text in relation to the text length (i.e. the number of characters of the text in the GT)
+
+Errors fall into one of the following three categories:
+
+* **deletion**: a character that is present in the text has been deleted from the output.
+
+Example:
+![A Fraktur sample reading "Sonnenfinſterniſſe:"](https://pad.gwdg.de/uploads/d7fa6f23-7c79-4fb2-ad94-7e98084c69d6.jpg)
+
+This reads `Sonnenfinſterniſſe:`. The output contains `Sonnenfinſterniſſe`, deleting `:`.
+
+* **substitution**: a character is replaced by another character in the output.
+
+Example:
+
+![A Fraktur sample reading "Die Finſterniſſe des 1801ſten Jahrs"](https://pad.gwdg.de/uploads/b894049b-8d98-4fe7-ac31-71b2c9393a6c.jpg)
+
+This heading reads `Die Finſterniſſe des 1801ſten Jahrs`. The output contains `180iſten`, replacing `1` with `i`.
+
+* **insertion**: a new character is introduced in the output.
+
+Example:
+
+![A Fraktur sample reading "diese Strahlen, und"](https://pad.gwdg.de/uploads/e6b6432e-d79c-4568-9aef-15a026c05b39.jpg)
+
+This reads `diese Strahlen, und`. The output contains `Strahlen ,`, inserting a white space before the comma.
+
+CER can be calculated in several ways, depending on whether a normalized CER is used or not.
+
+Given $i$ as the number of insertions, $d$ the number of deletions, $s$ the number of substitutions and $n$ the total number of characters in a text, the CER can be obtained by
+
+$CER = \frac{i + s+ d}{n}$
+
+If the CER value is calculated this way, it represents the percentage of characters incorrectly recognized by the OCR engine. Also, we can easily reach error rates beyond 100% when the output contains a lot of insertions.
+
+The *normalized* CER tries to mitigate this effect by considering the number of correct characters, $c$:
+
+$CER_n = \frac{i + s+ d}{i + s + d + c}$
+
+In OCR-D's benchmarking we calculate the *non-normalized* CER where values over 1 should be read as 100%.
```

Comment:
IIRC @cneud and @kba made this suggestion. Could one of you please elaborate?

Reply:
Just to bring this point from today's discussion to attention: Whether or not the CER denominator is normalized actually only matters if the segments which are aggregated/averaged are comparably small. If your method involves aligning pairs of linearised/concatenated **page texts**, then I agree we don't practically need this. But I would be surprised if that's what we want, because then our CER would entangle OCR quality with OLR/segmentation quality. 

(Of course, you can always argue: fine, so let's accompany CA/CER with FCA to be able to quantify that very difference. But I'd find it more natural/intuitive to build CA/CER by aggregating/averaging over **line pairs**, and **then** accompany this pure OCR metric by some pure OLR metric. OCR models are usually evaluated like this, so why shouldn't OCR workflows, too?)

Reply:
In terms of specifications one should be very careful with phrases like _usually_ or _common_ or alike that *might* transport hidden assumptions or presuppositions. We target different use cases. For example, rather small institutions that cannot afford training are probably less interested in the reduction of error rates, they just want means to 
a) get an idea what quality their current OCR is for reporting, or
b) want to define and assure a certain quality when assigning OCR-generation to an external service provider

From what I've read so far the emphasize on line-like evaluation is rather common in the context of model-training or fine-tuning since training components like Tesseract's-LSTM rely on line-text-pairs. From the perspective of QA without focus on training there are works which start evaluating at 1.000 chars (maybe because the DFG-Richtlinien promoted statistical evaluation methods like Bernoulli-XP, which introduced this number) or work even on complete corpora with linguistical or statistical approaches.

I'd suggest that it should be made clear what the basis for a single evaluation item is - a char, a word, a line, a page, a certain sample size (1.000+ chars for example), a corpus/subcorpus, or whatever. For example the term *WER* *might* suggest, that it's common base unit is a word, but at least we do not evaluate at this granularity, nor at line level. We aggregate upwards from frames or pages to higher units like years, newspapers or languages until we are at the very top. Thus we issue statements like: "According to our evaluation corpus with 1.000.000 Letters the Letter Accuracy (or Error rate, if you prefer) of our OCR-System is this or that." 

This wasn't clear to me until this week, and many of my irritations went off having this pointed out.

For the use case of VD-Historical prints (without newspapers) I'd vote for a single page as evaluation basis when going for textual metrics (CER, LER, WER, BoW). Maybe there can be some additional switch between lines/region/printspace to please the training community. 
Concerning Layout probably lines can make a good start point. Possibly this also holds for Fonts.

Reply:
@kba, @cneud and I came to a common understanding that **page** is the most suitable unit for evaluation in our case.

As @M3ssman mentioned line is the level to go to when re-training models, but as far as we're aware page level is most common (both for evaluation and layout).

Reply:
```suggestion
In OCR-D's benchmarking we calculate the *normalized* CER where values naturally range between 0 and 100%.
```

> @kba, @cneud and I came to a common understanding that **page** is the most suitable unit for evaluation in our case.
> 
> As @M3ssman mentioned line is the level to go to when re-training models, but as far as we're aware page level is most common (both for evaluation and layout).

Ok, fair enough. Then you should make that clear very early on – ideally at the top of the section (`Text Evaluation`), where it now just says `In OCR, one of these strings is generally the Ground Truth text and the other the recognized text which is the result of an OCR.` – this should then state directly that the text is concatenated at page level from smaller constituents in reading order.

(Perhaps you can even make a forward reference to the definition of FCA, which now explains this as `if applied naively by comparing concatenated page-level texts`.)